gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java

```diff
@@ -0,0 +1,136 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.gravitee.am.service.impl;
+
+import io.gravitee.am.model.ReferenceType;
+import io.gravitee.am.model.UserActivity;
+import io.gravitee.am.model.UserActivity.Type;
+import io.gravitee.am.repository.management.api.UserActivityRepository;
+import io.gravitee.am.service.UserActivityService;
+import io.gravitee.am.service.impl.user.activity.configuration.UserActivityConfiguration;
+import io.reactivex.Completable;
+import io.reactivex.Flowable;
+import io.reactivex.Single;
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import static io.gravitee.am.service.impl.user.activity.utils.CoordinateUtils.computeCoordinate;
+import static io.gravitee.am.service.impl.user.activity.utils.HashedKeyUtils.computeHash;
+
+/**
+ * @author Rémi SULTAN (remi.sultan at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@Component
+public class UserActivityServiceImpl implements UserActivityService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UserActivityServiceImpl.class);
+
+    private static final String LONGITUDE_KEY = "lon";
+    private static final String LATITUDE_KEY = "lat";
+    private static final String USER_AGENT = "user_agent";
+    private static final String LOGIN_ATTEMPTS = "login_attempts";
+
+    private final UserActivityConfiguration configuration;
+    private final UserActivityRepository userActivityRepository;
+
+    public UserActivityServiceImpl(
+            UserActivityConfiguration configuration,
+            @Lazy UserActivityRepository userActivityRepository
+    ) {
+        this.configuration = configuration;
+        this.userActivityRepository = userActivityRepository;
+    }
+
+    public boolean canSaveUserActivity() {
+        return this.configuration.isEnabled();
+    }
+
+    public long getRetentionTime() {
+        return this.configuration.getRetentionTime();
+    }
+
+    public ChronoUnit getRetentionUnit() {
+        return this.configuration.getRetentionUnit();
+    }
+
+    public Flowable<UserActivity> findByDomainAndTypeAndUserAndLimit(String domain, Type type, String userId, int limit) {
+        return userActivityRepository.findByDomainAndTypeAndKeyAndLimit(domain, type, buildKey(userId), limit);
+    }
+
+    public Completable save(
+            String domain,
+            String userId,
+            UserActivity.Type type,
+            Map<String, Object> data) {
+        final Date createdAt = new Date();
+        return Single.defer(() -> Single.just(new UserActivity()
```

Comment:
Why are we using a defer here to create a Bean ?

Reply:
We're building the hashed key of a user so we might want to defer in order to compute the hashed key when needed